### PR TITLE
feat: improve and complete `alloy` prelude crate feature flag compatiblity

### LINKED ISSUE #391: [Feature] Improve and complete `alloy` prelude crate feature flag compatiblity
### Component

provider, pubsub, rpc

### Describe the feature you would like

In order to allow users to use the `alloy` namespace to access all features we need to make sure there is full feature flag coverage of crates and we maintain that coverage. Another goal is to reduce the amount of individual feature flags by having a few high level feature flags which enable general code paths alongside individual specific feature flags.

Concretely we are currently missing the following feature flag exports:

- `alloy-rpc-client`:
  - `reqwest`
  - `hyper`
  - `pubsub`
  - `ws`
  - `ipc`
  
- `alloy-provider`
  - `provider-ws` misses `alloy-provider?/pubsub` flag
  - `provider-ipc` misses `alloy-provider?/pubsub` flag

- `alloy/core`
  - `native-keccak`

Other issues to resolve:

- Default to `std` as a default feature but support `no_std`
- Export missing `serde` feature flag
- Export all crates from `alloy/core`
- Refactor, clean up and regroup of feature flags

### Additional context

Related PR: #421

### PR DESCRIPTION
**Pull request by zerosnacks**
## Motivation

Closes #391, has a mirror PR for `alloy/examples` here: https://github.com/alloy-rs/examples/pull/39

## Solution

Fixes:
- Enables the `pubsub` feature on `alloy-provider` when using `provider-ws` or `provider-ipc`
- Exports `rpc-client-ws`, `rpc-client-ipc` as feature flags.
- Added missing `native-keccak` flag that was not re-exported from `alloy/core`.
- Default to `std` as a default feature but support `no_std`
- Add `reqwest` and `hyper` as top-level definitions and apply to `rpc-client` and `transport`
- Default to `reqwest` as default feature
- Export missing `serde` feature flag
- Reorder, clean up and regroup of feature flags

## PR Checklist

- [ ] Added Tests
- [ ] Added Documentation
- [x] Breaking changes

### CODE REVIEW

File: crates/alloy/Cargo.toml [hunk 1]
```diff
@@ -40,96 +49,128 @@ alloy-signer-wallet = { workspace = true, default-features = false, optional = t
...
+    "alloy-signer-wallet?/mnemonic-all-languages",
+]
+signer-yubihsm = ["signer-wallet", "alloy-signer-wallet?/yubihsm"]
 
-## transports
+# transports
 transports = ["dep:alloy-transport"]
 transport-http = ["transports", "dep:alloy-transport-http"]
-transport-http-reqwest = ["dep:reqwest", "alloy-transport-http?/reqwest"]
-transport-http-hyper = ["alloy-transport-http?/hyper"]
```
Comment (zerosnacks) [inline crates/alloy/Cargo.toml:132]: breaking change

Comment (zerosnacks) [reply crates/alloy/Cargo.toml:132]: Optionally we could mark these as deprecated instead of removing them outright

**Review by mattsse — APPROVED** [review]
all of these make sense to me!

File: crates/alloy/Cargo.toml [hunk 2]
```diff
@@ -41,95 +50,127 @@ alloy-signer-wallet = { workspace = true, default-features = false, optional = t
...
+    "dep:reqwest",
+    "alloy-rpc-client?/reqwest",
+    "alloy-transport-http?/reqwest",
+    "alloy-transport-http?/reqwest-default-tls",
 ]
+hyper = ["dep:hyper", "alloy-rpc-client?/hyper", "alloy-transport-http?/hyper"]
 
-# alloy
-default-tls = ["reqwest?/default-tls"]
-jsonrpsee-types = ["alloy-rpc-types?/jsonrpsee-types", "alloy-rpc-engine-types?/jsonrpsee-types"]
```
Comment (zerosnacks) [inline crates/alloy/Cargo.toml:85]: breaking change, renamed to `rpc-types-json`

File: crates/alloy/Cargo.toml [hunk 3]
```diff
@@ -41,95 +50,127 @@ alloy-signer-wallet = { workspace = true, default-features = false, optional = t
...
+reqwest = [
+    "dep:reqwest",
+    "alloy-rpc-client?/reqwest",
+    "alloy-transport-http?/reqwest",
+    "alloy-transport-http?/reqwest-default-tls",
 ]
+hyper = ["dep:hyper", "alloy-rpc-client?/hyper", "alloy-transport-http?/hyper"]
 
-# alloy
-default-tls = ["reqwest?/default-tls"]
```
Comment (zerosnacks) [inline crates/alloy/Cargo.toml:84]: breaking change, now part of `reqwest` flag

**Comment by prestwich** [comment]
how will this interact with #379? I'd prefer to merge that one first  as it is smaller

**Comment by zerosnacks** [comment]
> how will this interact with #379? I'd prefer to merge that one first as it is smaller

I've pulled in the latest changes, there were no breaking changes.

I've added the `alloy-provider` `reqwest` and `hyper` flag into the top level `reqwest` and `hyper` definition.

File: crates/alloy/Cargo.toml [hunk 4]
```diff
@@ -12,24 +12,33 @@ repository.workspace = true
 exclude.workspace = true
 
 [dependencies]
-alloy-core = { workspace = true, default-features = false, features = ["std"] }
+# core
+alloy-core = { workspace = true, default-features = false, optional = false }
```
Comment (prestwich) [inline crates/alloy/Cargo.toml:16]: `optional = false` is the default value and can be omitted

**Review by prestwich — CHANGES_REQUESTED** [review]
nit

File: crates/alloy/Cargo.toml (hunk 4 above)
Comment (zerosnacks) [reply crates/alloy/Cargo.toml:16]: fixed

**Pull request merged**